### PR TITLE
Enabled GPU capability, includes pipeline to optimally select hosts, and includes retry logic

### PR DESCRIPTION
**Pull request by harshp8l**

### CODE REVIEW

**Comment by EleanorRigby** [comment]
I think you will need to add host.rb also to this PR.

**Comment by harshp8l** [comment]
Included the host resource file

File: src/vsphere_cpi/lib/cloud/vsphere/resources/host.rb [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+      # @!attribute pci_devies
+      #   @return [Array<Vim::Host::PciDevice>] List of PCI Devices attached to host.
+      attr_accessor :pci_devices
+
+      # @!attribute graphics_info
+      #   @return [Array<Vim::Host::GraphicsInfo>] List of graphics devices attached to Host.
+      attr_accessor :graphics_info
+
+      # @!attribute runtime
+      #   @return [<Vim::Host::RuntimeInfo>] Runtime info for this host.
```
Comment (EleanorRigby) [inline src/vsphere_cpi/lib/cloud/vsphere/resources/host.rb]: Align @return with top

File: src/vsphere_cpi/lib/cloud/vsphere/resources/host.rb [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+          )
+        end
+      end
+
+      # @!attribute name
+      #   @return [String] host name.
+      attr_accessor :name
+
+      # @!attribute mob
+      #   @return [Vim::HostSystem] Host vSphere MOB.
```
Comment (EleanorRigby) [inline src/vsphere_cpi/lib/cloud/vsphere/resources/host.rb]: Align @return

Comment (EleanorRigby) [reply src/vsphere_cpi/lib/cloud/vsphere/resources/host.rb]: For all subsequent @return in this file too

File: src/vsphere_cpi/lib/cloud/vsphere/resources/host.rb [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+          mob.vm.any? do |vm|
+            begin
+              vm.config.hardware.device.any? do |device|
+                backing = device.backing
+                backing && backing.is_a?(VimSdk::Vim::Vm::Device::VirtualPCIPassthrough::DeviceBackingInfo) &&
+                  backing.id == gpu.pci_id
+              end
+            # This rescue is for VMs which are being created or being deleted by other CPI processes.
+            # VM in process of creation or deletion might throw up an error while we try to query its state and properties.
+            # Can we rescue more specific error here.
```
Comment (EleanorRigby) [inline src/vsphere_cpi/lib/cloud/vsphere/resources/host.rb]: Change this line to a TODO
`# TODO: Can we rescue ... ?`

File: src/vsphere_cpi/spec/unit/cloud/vsphere/resources/host_spec.rb [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+    end
+  end
+
+  describe '#graphics_info' do
+    it 'returns the list of graphics devices' do
+      expect(subject.graphics_info).to eq([1,2,3])
+    end
+  end
+
+  describe '#inspect' do
```
Comment (EleanorRigby) [inline src/vsphere_cpi/spec/unit/cloud/vsphere/resources/host_spec.rb]: Remove `inspect` from the unit test.

**Comment by nehagjain15** [comment]
Reopened this PR by cloning it into another branch on main repository

**Pull request closed without merging**